gc-worker/server/server.go

```diff
@@ -0,0 +1,299 @@
+	maxCallMsgSize := []grpc.DialOption{
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)),
+		grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(maxMsgSize)),
+	}
+	pdClient, err := pd.NewClientWithContext(
+		ctx, addrs, securityOption,
+		pd.WithGRPCDialOptions(maxCallMsgSize...),
+		pd.WithCustomTimeoutOption(10*time.Second),
+		pd.WithMaxErrorRetry(3),
+	)
+	if err != nil {
+		log.Error("fail to create pd client", zap.Error(err))
+		return errors.Trace(err)
+	}
+	s.pdClient = pdClient
+	return nil
+}
+
+func (s *Server) createEtcdClient() error {
+	endpoints := strings.Split(s.cfg.EtcdEndpoint, ",")
+	log.Info("create etcd v3 client", zap.Strings("endpoints", endpoints), zap.Reflect("cert", s.cfg.Security))
+
+	lgc := zap.NewProductionConfig()
+	lgc.Encoding = log.ZapEncodingName
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: etcdTimeout,
+		TLS:         nil,
+		LogConfig:   &lgc,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	s.etcdClient = client
+	return nil
+}
+
+// Close closes the server.
+func (s *Server) Close() {
+	if s.etcdClient != nil {
+		if err := s.etcdClient.Close(); err != nil {
+			log.Error("close etcd client meet error", zap.Error(err))
+		}
+	}
+	if s.pdClient != nil {
+		s.pdClient.Close()
+	}
+
+	if !atomic.CompareAndSwapInt64(&s.isServing, 1, 0) {
+		// server is already closed
+		return
+	}
+
+	log.Info("closing server")
+	s.stopServerLoop()
+
+	log.Info("closed server")
+}
+
+// checks whether server is closed or not.
+func (s *Server) IsServing() bool {
+	return atomic.LoadInt64(&s.isServing) != 0
+}
+
+func (s *Server) IsLead() bool {
+	return atomic.LoadInt64(&s.isLead) != 0
+}
+
+func (s *Server) StartServer() {
+	atomic.StoreInt64(&s.isServing, 1)
+	s.serverLoopCtx, s.serverLoopCancel = context.WithCancel(s.ctx)
+	s.serverLoopWg.Add(2)
+	go s.startEtcdLoop()
+	go s.startUpdateGCSafePointLoop()
+}
+
+func (s *Server) stopServerLoop() {
+	s.serverLoopCancel()
+	s.serverLoopWg.Wait()
+}
+
+// ComposeTS creates a ts from physical and logical parts.
+func ComposeTS(physical, logical int64) uint64 {
+	return uint64((physical << physicalShiftBits) + logical)
+}
+
+// ExtractPhysical returns a ts's physical part.
+func ExtractPhysical(ts uint64) int64 {
+	return int64(ts >> physicalShiftBits)
+}
+
+// GetPhysical returns physical from an instant time with millisecond precision.
+func GetPhysical(t time.Time) int64 {
+	return t.UnixNano() / int64(time.Millisecond)
+}
+
+// GetTimeFromTS extracts time.Time from a timestamp.
+func GetTimeFromTS(ts uint64) time.Time {
+	ms := ExtractPhysical(ts)
+	return time.Unix(ms/1e3, (ms%1e3)*1e6)
+}
+
+func (s *Server) startEtcdLoop() {
+	defer s.serverLoopWg.Done()
+	ctx, cancel := context.WithCancel(s.serverLoopCtx)
+	defer cancel()
+
+	for {
+		select {
+		case <-time.After(etcdCampaignInterval):
+			session, err := concurrency.NewSession(s.etcdClient, concurrency.WithTTL(5))
```

Comment:
Do we need to check leadership before TTL expired ?
I think there should be some cases that a leader will lose its leadership. E.g, network isolation, process hang up.
Maybe we should check leadership periodically.

Reply:
if the leader process/node crash or be killed, folllower node can get the leadership automatically.

Reply:
What will happen in the following scenario, can worker0 know it is not leader anymore ?
1. Worker0 become leader
2. Network between worker0 & etcd broken
3. Worker1 become leader
4. Network between worker0 & etcd restored

Reply:
a channel reciever is added to recieve the `session.Done` signal. If `Network between worker0 & etcd broken` occurs, the session should be `Done`, and the `leader flag` can be set to `false`. I have not test this yet.

Reply:
OK. I think we can leave it over now, and come back after some tests.

p.s. I look through leadership codes of PD and see there are logics for ["keey leader"](https://github.com/tikv/pd/blob/0734f98770950fa1f88edf3aeb43c522bb429b35/server/server.go#L1320) and ["check leader"](https://github.com/tikv/pd/blob/0734f98770950fa1f88edf3aeb43c522bb429b35/server/server.go#L1380). We can refer to them.